source/ShopPanel.cpp

```diff
@@ -562,29 +430,113 @@ bool ShopPanel::KeyDown(SDL_Keycode key, Uint16 mod, const Command &command, boo
+		// Outfitter: Buy and install up to <modifier> outfits for each selected ship.
+		// Shipyard: Buy up to <modifier> ships.
+		result = CanDoBuyButton();
 		if(result)
+			DoBuyButton();
+	}
+	else if(key == 's')
+	{
+		// In the Shipyard: Sell selected ships and outfits.
+		// In the Outfitter: Sell <modifier> of the selected outfit from each selected ship.
+		if(isOutfitter)
 		{
-			DoTransaction(key);
-
-			// Ship-based updates to cargo are handled when leaving.
-			// Ship-based selection changes are asynchronous, and handled by ShipyardPanel.
-			if(isOutfitter)
+			result = CanSellOrUninstall("sell");
+		}
+		else
+		{
+			result = playerShip;
+		}
+		if(result)
+			Sell(false);
+	}
+	else if(key == 'r')
+	{
+		// In the Shipyard: Sell selected ships and move outfits to Storage.
+		// In the Outfitter: Move <modifier> of the selected outfit to storage from either cargo (first) or else each
+		// of the selected ships.
+		if(isOutfitter)
+		{
+			result = CanSellOrUninstall("store");
+			if(result)
+				RetainInStorage();
+		}
+		else if(playerShip)
+		{
+			Sell(true);
+		}
+	}
+	else if(key == 'c' && isOutfitter)
+	{
+		// Either move up to <multiple> outfits into cargo from storage if any are in storage, or else buy up to
+		// <modifier> outfits into cargo.
+		// Note: If the outfit is not able to be moved from storage or bought into cargo, give an error based on the buy
+		// condition.
+		if(CanMoveToCargoFromStorage())
+		{
+			MoveToCargoFromStorage();
+		}
+		else
+		{
+			// Selected outfit cannot be moved from storage, try buying:
+			result = CanBuyToCargo();
+			if(result)
+				BuyIntoCargo();
+		}
+	}
+	else if(key == 'i' && isOutfitter)
+	{
+		// Install up to <modifier> outfits from already owned equipment into each selected ship.
+		result = CanInstall();
+		if(result)
+			Install();
+	}
+	else if(key == 'u')
+	{
+		// In the Shipyard: Sell selected ships and move outfits to Storage (previous behavior).
+		// In the Outfitter: Move <modifier> of the selected outfit to storage from each selected ship or from cargo.
+		if(isOutfitter)
+		{
+			// Uninstall up to <multiple> outfits from each of the selected ships if any are available to uninstall, or
+			// else move up to <multiple> outfits from cargo into storage.
+			// Note: If the outfit is not able to be uninstalled or moved from cargo, give an error based on the
+			// uninstall condition.
+			result = CanSellOrUninstall("uninstall");
+			if(result)
 			{
-				player.UpdateCargoCapacities();
-				CheckSelection();
+				Uninstall();
+			}
+			else if(CanSellOrUninstall("store"))
+			{
+				RetainInStorage();
+				result = true;
 			}
 		}
-		else if(result.HasMessage())
-			GetUI()->Push(new Dialog(result.Message()));
+		else if(playerShip)
+		{
+			// Shipyard, old behavior, treat 'u' the same as 'r': Sell ship and retain the outfits in storage.
+			Sell(true);
+		}
 	}
 	else
+	{
 		return false;
+	}
+
+	if(result.HasMessage())
```

Comment:
If there's an error, we'll show it here.  Otherwise, everything went well and so if we're in the Outfitter, we need to apply the updates to Cargo as before.